pydantic/_internal/_known_annotated_metadata.py

```diff
@@ -222,142 +231,39 @@ def apply_known_metadata(annotation: Any, schema: CoreSchema) -> CoreSchema | No
-                schema,
-            )
-        elif constraint == 'min_length':
-            s = cs.no_info_after_validator_function(
-                partial(_validators.min_length_validator, min_length=value),
-                schema,
-            )
-            if _is_sequence_metadata(schema):
-                add_js_update_schema(s, lambda: {'minItems': (as_jsonable_value(value))})
-                return s
-            add_js_update_schema(s, lambda: {'minLength': (as_jsonable_value(value))})
-            return s
-        elif constraint == 'max_length':
-            s = cs.no_info_after_validator_function(
-                partial(_validators.max_length_validator, max_length=value),
-                schema,
-            )
-            if _is_sequence_metadata(schema):
-                add_js_update_schema(s, lambda: {'maxItems': (as_jsonable_value(value))})
-                return s
-            add_js_update_schema(s, lambda: {'maxLength': (as_jsonable_value(value))})
-            return s
-        elif constraint == 'strip_whitespace':
-            return cs.chain_schema(
-                [
-                    schema,
-                    cs.str_schema(strip_whitespace=True),
-                ]
-            )
-        elif constraint == 'to_lower':
-            return cs.chain_schema(
-                [
-                    schema,
-                    cs.str_schema(to_lower=True),
-                ]
-            )
-        elif constraint == 'to_upper':
-            return cs.chain_schema(
-                [
-                    schema,
-                    cs.str_schema(to_upper=True),
-                ]
-            )
-        elif constraint == 'min_length':
-            return cs.no_info_after_validator_function(
-                partial(_validators.min_length_validator, min_length=annotation.min_length),
-                schema,
+        if constraint in {'pattern', 'strip_whitespace', 'to_lower', 'to_upper', 'coerce_numbers_to_str'}:
+            chain_schema_steps.append(cs.str_schema(**{constraint: value}))
+        elif constraint in {'gt', 'ge', 'lt', 'le', 'multiple_of', 'min_length', 'max_length'}:
+            if constraint == 'multiple_of':
+                json_schema_constraint = 'multiple_of'
+            elif constraint in {'min_length', 'max_length'}:
+                if schema['type'] == 'list' or (
+                    schema['type'] == 'json-or-python' and schema['json_schema']['type'] == 'list'
+                ):
+                    json_schema_constraint = 'minItems' if constraint == 'min_length' else 'maxItems'
+                else:
+                    json_schema_constraint = 'minLength' if constraint == 'min_length' else 'maxLength'
+            else:
+                json_schema_constraint = constraint
+
+            schema = cs.no_info_after_validator_function(
+                partial(COMPARISON_VALIDATORS[constraint], **{constraint: value}), schema
             )
```

Comment:
This might be slightly less readable, but I'd argue the redundancy before was worse...